index.bs

```diff
@@ -3763,10 +3772,656 @@ nothrow>WritableStreamDefaultControllerError ( <var>controller</var>, <var>error
+{{ReadableStream}} constructor when constructing the \[[readable]] slot.
+
+<div class="note">
+  This specification uses the public API for transmitting events from the {{ReadableStream}} back to the
+  {{TransformStream}} implementation for simplicity. Since the {{TransformStreamDefaultSource}} class is not observable,
+  implementations do not need to implement it, but could instead do something else observably equivalent. See <a
+  href="https://github.com/whatwg/streams/issues/813">#813</a> for an alternate specification approach that will make
+  this clearer in the future.
+</div>
+
+<h4 id="ts-default-source-class-definition">Class Definition</h4>
+
+<div class="non-normative">
+
+<em>This section is non-normative.</em>
+
+If one were to write the {{TransformStreamDefaultSource}} class in something close to the syntax of [[!ECMASCRIPT]],
+it would look like
+
+<pre><code class="lang-javascript">
+  class TransformStreamDefaultSource {
+    constructor(stream, startPromise)
+
+    start()
+    pull()
+    cancel(reason)
+  }
+</code></pre>
+
+</div>
+
+<h4 id="ts-default-source-internal-slots">Internal Slots</h4>
+
+Instances of {{TransformStreamDefaultSource}} are created with the internal slots described in the following table:
+
+<table>
+  <thead>
+    <tr>
+      <th>Internal Slot</th>
+      <th>Description (<em>non-normative</em>)</th>
+    </tr>
+  </thead>
+  <tr>
+    <td>\[[ownerTransformStream]]
+    <td class="non-normative">The {{TransformStream}} instance
+  </tr>
+  <tr>
+    <td>\[[startPromise]]
+    <td class="non-normative">The <var>startPromise</var> parameter that was passed to the constructor
+  </tr>
+</table>
+
+<h4 id="ts-default-source-constructor" constructor for="TransformStreamDefaultSource"
+lt="TransformStreamDefaultSource(stream, startPromise)">new TransformStreamDefaultSource(<var>stream</var>, <var>startPromise</var>)</h4>
+
+<emu-alg>
+  1. Set *this*.[[ownerTransformStream]] to _stream_.
+  1. Set *this*.[[startPromise]] to _startPromise_.
+</emu-alg>
+
+<h4 id="ts-default-source-prototype">Properties of the {{TransformStreamDefaultSource}} Prototype</h4>
+
+<h5 id="ts-default-source-start" method for="TransformStreamDefaultSource">start()</h5>
+
+<emu-alg>
+  1. Return *this*.[[startPromise]].
+</emu-alg>
+
+<h5 id="ts-default-source-pull" method for="TransformStreamDefaultSource">pull()</h5>
+
+<emu-alg>
+  1. Let _stream_ be *this*.[[ownerTransformStream]].
+  1. Assert: _stream_.[[backpressure]] is *true*.
+  1. Assert: _stream_.[[backpressureChangePromise]] is not *undefined*.
+  1. Perform ! TransformStreamSetBackpressure(_stream_, *false*).
+  1. Return _stream_.[[backpressureChangePromise]].
+</emu-alg>
+
+<h5 id="ts-default-source-cancel" method for="TransformStreamDefaultSource">cancel(<var>reason</var>)</h5>
+
+<emu-alg>
+  1. Perform ! TransformStreamError(*this*.[[ownerTransformStream]], _reason_).
```

Comment:
Just noticing this... is this correct behavior? I'd have thought it'd be more like terminate than error.

Reply:
It actually does behave like terminate, because the readable side is already closed when cancel() is called. I added a note to make this clearer.

Reply:
Is there a reason to call TransformStreamError instead of TransformStreamDefaultControllerTerminate? I guess layering, hmm... seems to argue for factoring out a TransformStreamTerminate, if I am understanding correctly.

Reply:
We'd end up with TypeError('TransformStream terminated') as the storedError in the writable. This seems less desirable to me than retaining whatever was passed to cancel().

However... maybe factoring out the "bottom half" of TransformStreamDefaultControllerTerminate and TransformStreamError is the way to go. I will try it.

Reply:
I made factored out a new abstract op named "TransformStreamErrorWritableAndUnblockWrite". PTAL.